### DISCUSSION TITLE
Merging pull requests with forge

### DISCUSSION THREAD

**Discussion by Mastergamer433** [magit/forge#524] [Support] [answered]
I can't seem to find the binding in the magit transient ui for merging a pull request with forge, so i have to use `M-x forge-merge` for that.

And btw the box down that says "I have done a search or similar discussions" I have done 3 searches for similar discussions so.

**Comment by tarsius** [comment, marked as answer]
I recommend you do not use `forge-merge`, as it performs the merge using the Github API, and doing that results in inferior merges, depending on which "merge method" you choose. Github does not offer basic "fast-forward" merge, which is extremely annoying. This command exists for users whose employer forces them to use the API to perform merges.

To perform a merge in Emacs, use Magit's merge commands. You can, for example, place the cursor on a pr in the pull-request list in the status buffer and type `m m`.

What I usually do is to check out the pr using ~~`b n`~~ `b f` on the pr in the status buffer, then (possibly after *) I merge the checked out branch into to proper branch (likely `main`/`master`) using `m i RET`. Finally I push to `origin/{main/master}`, at which point Github becomes aware that the merge has happened, so it automatically closes the pr.

At * I might make changes to the commits in the pr. If I do, then I push to the contributors feature branch *before* merging into the target branch (or rather before *pushing* the result). Doing that ensures that Github later realizes that the merge has happened.

**Comment by Mastergamer433** [comment]
So I should checkout that branch that is source in the pr (place the cursor on pr and click `b n`) then press `m i RET` then `p o upstream/main` and then its finnished?

**Reply by tarsius** [reply]
Yes.

Or (with `master` checked out already and point on the pr) `m m RET`, followed by `p o RET`.

But I admit there is room for improvement here, especially concerning the automatic commit message (when actually creating a merge commit, which I usually don't do). For me the checkout, merge, push workflow works very well, but there is room for improvement when using another path.

In general my recommendation boils down to: use Magit (there are many more workflows available), then push the result.

There is no need for Forge-specific commands, but there is always room to improve Magit's commands (and when doing that, the needs of users who also use Forge should of course be taken into account).

**Reply by Mastergamer433** [reply]
When i press `b n` i get the create branch thing so why would i use `b n`? That does not checkout any branch it just creates one.

So am I doing it wrong or what should i do? 

I place my cursor on the pr and the click `b n` and it prompts for a name for the new branch so yeah that does not work.

**Reply by tarsius** [reply]
Ah, that's a personal binding. It was `b n` for a short while, but I had to revert that, see 7157b6ee994d85b74803ee2a8c98ce1bf17df8bf. However in my own config I stuck to `n`.

The default binding is `b f`.

**Reply by Mastergamer433** [reply]
While trying to checkout the branch it says:
```
magit-process-finish: Git error: "Could not fetch Mastergamer433 (in /home/mg433/Projects/websiteBackend/)"
```
and i did `create new "pr-12" branch instead` when checking out that branch.

**Comment by tarsius** [comment]
The fetching happens after the api was used to perform the merge to bring the local state up-to-date. That shouldn't affect the outcome of the merge (visit the webpage to confirm) and there are a variety of reasons why it might fail.

The fetching is done using `forge-pull` (same as `f n`). Try that command separately to see whether that fails too. Maybe it doesn't and the issue was just some temporary hickup on the server side. Or maybe the remote is misconfigured somehow on your end.
